src/controller_examples/zookeeper_controller/proof/common.rs

```diff
@@ -130,252 +39,79 @@ pub open spec fn at_zookeeper_step(key: ObjectRef, step: ZookeeperReconcileStep)
         &&& s.reconcile_state_of(zk.object_ref()).pending_req_msg.is_Some()
         &&& s.message_in_flight(s.pending_req_of(zk.object_ref()))
-        &&& is_create_config_map_request_msg(s.pending_req_of(zk.object_ref()), zk)
+        &&& is_correct_pending_request_msg_at_zookeeper_step(step, s.pending_req_of(zk.object_ref()), zk, object)
     }
 }
 
-pub open spec fn req_msg_is_the_in_flight_pending_req_at_after_create_config_map_step_with_zk(
-    zk: ZookeeperClusterView, req_msg: Message
+pub open spec fn req_msg_is_the_in_flight_pending_req_at_zookeeper_step_with_zk(
+    step: ZookeeperReconcileStep, zk: ZookeeperClusterView, req_msg: Message, object: DynamicObjectView
 ) -> StatePred<ClusterState> {
     |s: ClusterState| {
-        &&& at_after_create_config_map_step_with_zk(zk)(s)
+        &&& at_zookeeper_step_with_zk(zk, step)(s)
         &&& s.reconcile_state_of(zk.object_ref()).pending_req_msg == Option::Some(req_msg)
         &&& s.message_in_flight(req_msg)
-        &&& is_create_config_map_request_msg(req_msg, zk)
+        &&& is_correct_pending_request_msg_at_zookeeper_step(step, req_msg, zk, object)
     }
 }
 
-pub open spec fn at_after_create_config_map_step_with_zk_and_exists_resp_in_flight(
-    zk: ZookeeperClusterView
+pub open spec fn exists_resp_in_flight_at_zookeeper_step_with_zk(
+    step: ZookeeperReconcileStep, zk: ZookeeperClusterView, object: DynamicObjectView
 ) -> StatePred<ClusterState> {
     |s: ClusterState| {
-        &&& at_after_create_config_map_step_with_zk(zk)(s)
+        &&& at_zookeeper_step_with_zk(zk, step)(s)
         &&& s.reconcile_state_of(zk.object_ref()).pending_req_msg.is_Some()
-        &&& is_create_config_map_request_msg(s.pending_req_of(zk.object_ref()), zk)
+        &&& is_correct_pending_request_msg_at_zookeeper_step(step, s.pending_req_of(zk.object_ref()), zk, object)
         &&& exists |resp_msg| {
             &&& #[trigger] s.message_in_flight(resp_msg)
             &&& resp_msg_matches_req_msg(resp_msg, s.pending_req_of(zk.object_ref()))
         }
     }
 }
 
-pub open spec fn resp_msg_is_the_in_flight_resp_at_after_create_config_map_step_with_zk(
-    zk: ZookeeperClusterView, resp_msg: Message
+pub open spec fn resp_msg_is_the_in_flight_resp_at_zookeeper_step_with_zk(
+    step: ZookeeperReconcileStep, zk: ZookeeperClusterView, resp_msg: Message, object: DynamicObjectView
 ) -> StatePred<ClusterState> {
     |s: ClusterState| {
-        &&& at_after_create_config_map_step_with_zk(zk)(s)
+        &&& at_zookeeper_step_with_zk(zk, step)(s)
         &&& s.reconcile_state_of(zk.object_ref()).pending_req_msg.is_Some()
-        &&& is_create_config_map_request_msg(s.pending_req_of(zk.object_ref()), zk)
+        &&& is_correct_pending_request_msg_at_zookeeper_step(step, s.pending_req_of(zk.object_ref()), zk, object)
         &&& s.message_in_flight(resp_msg)
         &&& resp_msg_matches_req_msg(resp_msg, s.pending_req_of(zk.object_ref()))
     }
 }
```

Comment:
I think the above spec functions can be made more generic and will be reused heavily. But to do so we might need to make the `step` an associated type of `Reconciler` and the spec functions need to take a closure for `is_correct_pending_request_msg_at_zookeeper_step`. We can work on this later.

Reply:
Absolutely.

Reply:
> take a closure for `is_correct_pending_request_msg_at_zookeeper_step`

Actually I have more thoughts about this... I will put it in the issue.